build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/tscTask.ts

```diff
@@ -258,13 +259,56 @@ export class TscTask extends LeafTask {
 			path.resolve,
 		);
 
-		if (!isEqual(configOptions, tsBuildInfoOptions)) {
-			this.traceTrigger(`ts option changed ${configFileFullPath}`);
-			this.traceTrigger("Config:");
-			this.traceTrigger(JSON.stringify(configOptions, undefined, 2));
-			this.traceTrigger("BuildInfo:");
-			this.traceTrigger(JSON.stringify(tsBuildInfoOptions, undefined, 2));
-			return false;
+		// The following code checks that the cached buildInfo options -- that is, the tsconfig options used to build
+		// originally -- match the current tsconfig. If they don't match, then the cache is outdated and a rebuild is
+		// needed.
+		//
+		// However, there are some tsconfig settings that don't seem to be cached in the tsbuildinfo. This makes any builds
+		// including these settings always fail this check. We special-case the properties in this set -- any differences in
+		// those fields are ignored.
+		//
+		// IMPORTANT: This is a somewhat unsafe action. Technically speaking, we don't know for sure the incremental build
+		// status when there is ANY difference between these settings. Thus the safest thing to do is to assume a rebuild is
+		// needed, Projects using these ignored settings may exhibit strange incremental build behavior.
+		//
+		// For that reason, this behavior can be disabled completely using the _FLUID_BUILD_DISABLE_IGNORE_TSC_OPTIONS_
+		// environment variable. If that variable is set to any non-empty value, the list of ignored options will not be
+		// checked.
+		const tsConfigOptionsIgnored: Set<string> =
+			(process.env._FLUID_BUILD_DISABLE_IGNORE_TSC_OPTIONS_ ?? "" !== "")
+				? new Set()
+				: new Set(["allowJs", "checkJs"]);
+
+		const diffResults = diffObjects(configOptions, tsBuildInfoOptions);
+		if (diffResults.length > 0) {
+			// Track whether the diff is "real"; that is, the diff is in properties that are not ignored. By default,
+			// assume that it is.
+			let diffIsReal = true;
+
+			// We only need to check the different properties if the number of differences is <= the ignored options;
+			// any properties over that count will be a non-ignored difference.
+			if (diffResults.length <= tsConfigOptionsIgnored.size) {
+				// The diff is "real" if any different property is not ignored
+				diffIsReal = diffResults.some((diffResult) => {
+					const isIgnoredOption = tsConfigOptionsIgnored.has(diffResult.path);
+					if (isIgnoredOption) {
+						this.traceTrigger(`ignoring tsbuildinfo property: ${diffResult.path}`);
+					}
+					return !isIgnoredOption;
+				});
+			}
+
+			if (diffIsReal) {
+				this.traceTrigger(`ts options changed ${configFileFullPath}`);
+				this.traceTrigger("Diff:");
+				this.traceTrigger(JSON.stringify(diffResults, undefined, 2));
```

Comment:
When there is a difference and ignored things are involved, it would be good to distinguish those items. Otherwise, there will likely be confusion about how to address the differences that can't be addressed.
It would be fine to just omit them here if the ignore check ran all the time because they would be listed as ignored. Recommend not optimizing and just processing the diffResult to remove those ignored.